### PR TITLE
kubeadm v1beta1: InitConfiguration.APIEndpoint -> LocalAPIEndpoint

### PR DESCRIPTION
**Pull request by luxas**
**What type of PR is this?**
> Uncomment only one, leave it on its own line:
>
/kind api-change
> /kind bug
> /kind cleanup
> /kind design
> /kind documentation
> /kind failing-test
> /kind feature
> /kind flake

**What this PR does / why we need it**:

As discussed in today's SIG meeting, `LocalAPIEndpoint` makes more sense than `APIEndpoint` which is more generic. The value is specific to the current master being initialized, and is different from ControlPlaneEndpoint which refers to the public endpoint for all API servers. This change targets the upcoming v1beta1 Config API.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
kubeadm v1beta1 API: InitConfiguration.APIEndpoint has been renamed to .LocalAPIEndpoint
```
@kubernetes/sig-cluster-lifecycle-pr-reviews

### CODE REVIEW

File: cmd/kubeadm/app/apis/kubeadm/types.go [hunk 1]
```diff
@@ -46,8 +46,13 @@ type InitConfiguration struct {
 	// NodeRegistration holds fields that relate to registering the new master node to the cluster
 	NodeRegistration NodeRegistrationOptions
 
-	// APIEndpoint represents the endpoint of the instance of the API server to be deployed on this node.
-	APIEndpoint APIEndpoint
+	// LocalAPIEndpoint represents the endpoint of the API server instance that's deployed on this control plane node
+	// In HA setups, this differ from ClusterConfiguration.ControlPlaneEndpoint in the sense that ControlPlaneEndpoint
```
Comment (timothysc) [inline cmd/kubeadm/app/apis/kubeadm/types.go]: differs 
or is different

**Review by timothysc — APPROVED** [review]
/lgtm 
/approve 

docs can be cleaned up later.

**Comment by timothysc** [comment]
/hold 
@fabriziopandini - in case you have any comments on the naming.

**Comment by neolit123** [comment]
should this be considered as "action required" item?

**Comment by fabriziopandini** [comment]
I'm fine with prefixing with Local if it helps with clarification.
Only wondering if we need to rename also `ClusterStatus.APIEndpoints` --> `ClusterStatus.LocalAPIEndpoints` for consistency

@timothysc @luxas @rosti @neolit123 opinions? quick thumbs up/down?

**Review by rosti — COMMENTED** [review]
@luxas thanks for this change! It's nice to have you back!

I do think, that we should actually extend this to cover `JoinConfiguration.APIEndpoint -> LocalAPIEndpoint` too. Although, that can be done in another PR.

@fabriziopandini IDK about `APIEndpoints` vs `LocalAPIEndpoints`. Both ways seem descriptive enough in its context.

File: cmd/kubeadm/app/apis/kubeadm/validation/validation.go [hunk 2]
```diff
@@ -49,7 +49,7 @@ func ValidateInitConfiguration(c *kubeadm.InitConfiguration) field.ErrorList {
 	allErrs = append(allErrs, ValidateNodeRegistrationOptions(&c.NodeRegistration, field.NewPath("nodeRegistration"))...)
 	allErrs = append(allErrs, ValidateBootstrapTokens(c.BootstrapTokens, field.NewPath("bootstrapTokens"))...)
 	allErrs = append(allErrs, ValidateClusterConfiguration(&c.ClusterConfiguration)...)
-	allErrs = append(allErrs, ValidateAPIEndpoint(&c.APIEndpoint, field.NewPath("apiEndpoint"))...)
+	allErrs = append(allErrs, ValidateAPIEndpoint(&c.LocalAPIEndpoint, field.NewPath("apiEndpoint"))...)
```
[the remark this answers is not in the document]
Comment (luxas) [reply cmd/kubeadm/app/apis/kubeadm/validation/validation.go]: done

File: cmd/kubeadm/app/apis/kubeadm/types.go (hunk 1 above)
Comment (luxas) [reply cmd/kubeadm/app/apis/kubeadm/types.go]: done

**Comment by luxas** [comment]
> I do think, that we should actually extend this to cover JoinConfiguration.APIEndpoint -> LocalAPIEndpoint

@fabriziopandini will implement this, along with changing `JoinConfiguration.ControlPlane` to be a pointer struct

**Comment by neolit123** [comment]
the verify failures could be due to golang 1.11.x.
@luxas latest golang with ./hack/update-gofmt should fix it.

**Comment by luxas** [comment]
@neolit123 yeah I rebased and fixed that in the latest push
Please re-apply the label

File: cmd/kubeadm/app/apis/kubeadm/v1beta1/types.go [hunk 3]
```diff
@@ -45,8 +45,13 @@ type InitConfiguration struct {
...
 
-	// APIEndpoint represents the endpoint of the instance of the API server to be deployed on this node.
-	APIEndpoint APIEndpoint `json:"apiEndpoint,omitempty"`
+	// LocalAPIEndpoint represents the endpoint of the API server instance that's deployed on this control plane node
+	// In HA setups, this differs from ClusterConfiguration.ControlPlaneEndpoint in the sense that ControlPlaneEndpoint
+	// is the global endpoint for the cluster, which then loadbalances the requests to each individual API server. This
+	// configuration object lets you customize what IP/DNS name and port the local API server advertises it's accessible
+	// on. By default, kubeadm tries to auto-detect the IP of the default interface and use that, but in case that process
+	// fails you may set the desired value here.
+	LocalAPIEndpoint APIEndpoint `json:"localAPIEndpoint,omitempty"`
```
Comment (liggitt) [inline cmd/kubeadm/app/apis/kubeadm/v1beta1/types.go]: Was the old field shipped and persisted previously? If so, what will happen if a previously serialized config file is read and the old apiEndpoint field is discarded and the new localAPIEndpoint field is empty?

**Review by liggitt — COMMENTED** [review]
Is this related to 
https://github.com/kubernetes/kubernetes/issues/70907 ?

**Comment by neolit123** [comment]
> Is this related to
#70907 ?

@liggitt we merged a e2e fix 11 hours ago. :)

File: cmd/kubeadm/app/apis/kubeadm/v1beta1/types.go (hunk 3 above)
Comment (neolit123) [reply cmd/kubeadm/app/apis/kubeadm/v1beta1/types.go]: yep, it was shipped as part of the previous config version v1alpha3.

if an old config YAML is read and the old field is found, it _should be_ converted to the new field.
https://github.com/kubernetes/kubernetes/pull/70761/files#diff-628349c40af3952df65127e8449c04e2R31

kubeadm 1.13 would support both v1alpha3 and v1beta1.

Comment (liggitt) [reply cmd/kubeadm/app/apis/kubeadm/v1beta1/types.go]: I specifically meant if v1beta1 was shipped in a previous release with this field. If another version of the config was shipped, but handles conversion, that's fine

Comment (neolit123) [reply cmd/kubeadm/app/apis/kubeadm/v1beta1/types.go]: yep, v1beta1 would be shipped for the first time in 1.13.

**Pull request merged**